packages/botonic-plugin-dialogflow/src/index.js

```diff
@@ -1,66 +1,102 @@
+import uuid from 'uuid'
 import axios from 'axios'
+import { KJUR } from 'jsrsasign'
 
 export default class BotonicPluginDialogflow {
-  constructor(options) {
-    this.options = options
-    this.defaultVersion = '20150910'
+  constructor(creds) {
+    this.projectID = creds.project_id
+    this.sessionID = uuid.v4()
+    this.creds = creds
+    this.token = null
   }
 
-  async query(data, lang, sessionId) {
-    // Data, lang, sessionId and version are required parameters
-    // See https://dialogflow.com/docs/reference/agent/query for more information
-    return await axios({
-      headers: {
-        Authorization: `Bearer ${this.options.token}`
-      },
-      url: `https://api.dialogflow.com/v1/query`,
-      params: {
-        query: data,
-        lang: lang,
-        sessionId: sessionId,
-        v: this.options.version || this.defaultVersion
-      }
-    })
+  async getToken() {
+    if (!this.token) {
+      this.token = await this.generateToken(this.creds)
+    }
+    return this.token
   }
 
   async pre({ input, session, lastRoutePath }) {
     try {
       await this.dialogflowToInput({ input, session, lastRoutePath })
     } catch (error) {
       console.log(error.response)
+      await this.refreshToken()
+      await this.dialogflowToInput({ input, session, lastRoutePath })
     }
     return { input, session, lastRoutePath }
   }
 
   async post({ input, session, lastRoutePath, response }) {}
 
+  async refreshToken() {
+    this.token = await this.generateToken(this.creds)
+  }
+
+  async generateToken(creds) {
+    const header = {
+      alg: 'RS256',
+      typ: 'JWT',
+      kid: creds.private_key_id
+    }
+    const payload = {
+      iss: creds.client_email,
+      sub: creds.client_email,
+      iat: KJUR.jws.IntDate.get('now'),
+      exp: KJUR.jws.IntDate.get('now + 1hour'),
+      aud:
+        'https://dialogflow.googleapis.com/google.cloud.dialogflow.v2.Sessions'
+    }
+
+    const stringHeader = JSON.stringify(header)
+    const stringPayload = JSON.stringify(payload)
+    let token = KJUR.jws.JWS.sign(
+      'RS256',
+      stringHeader,
+      stringPayload,
+      creds.private_key
+    )
+    return token
+  }
+
+  async query(queryData, languageCode = 'en-US') {
+    return axios({
+      method: 'post',
+      url: `https://dialogflow.googleapis.com/v2/projects/${this.projectID}/agent/sessions/${this.sessionID}:detectIntent`,
+      headers: {
+        Authorization: `Bearer ${await this.getToken()}`,
+        'Content-Type': 'application/json'
+      },
+      data: {
+        queryInput: {
+          text: {
+            text: queryData,
+            languageCode: languageCode
+          }
+        }
+      }
+    })
+  }
+
   async dialogflowToInput({ input, session, lastRoutePath }) {
-    let df_session_id = session.user.id
-    let intent = null
     let confidence = 0
+    let intent = null
```

Comment:
if these default values are never used, I would just assign them to their final values. eg.
`let entities = queryResult.parameters`